packages/auth/lib/challenges/EmailCodeChallenge.js

```diff
@@ -0,0 +1,82 @@
+const { sendMailTemplate } = require('@orbiting/backend-modules-mail')
+
+const { newAuthError } = require('../AuthError')
+const t = require('../t')
+
+const TokensExceededError = newAuthError('email-code-challenge-tokens-exceeded', 'api/auth/emailCode/tokensExceeded')
+const CollisionError = newAuthError('email-code-challenge-collision', 'api/auth/emailCode/collison')
+
+const CODE_LENGTH = 6
+const MAX_VALID_TOKENS = 5
+const TTL = 1000 * 60 * 60 // 60 minutes
+const Type = 'EMAIL_CODE'
+
+const { DEFAULT_MAIL_FROM_ADDRESS } = process.env
+
+module.exports = {
+  Type,
+  generateNewToken: async ({ email, pgdb }) => {
+    const existingPayloads = (await pgdb.public.tokens.find(
+      {
+        type: Type,
+        email,
+        'expiresAt >=': new Date()
+      }
+    ))
+      .map(token => token.payload)
+
+    if (existingPayloads.length >= MAX_VALID_TOKENS) {
+      console.error('Unable to generate a new token: Found too many valid tokens.')
+      throw new TokensExceededError({ email, tokens: existingPayloads, MAX_VALID_TOKENS })
+    }
+
+    let attempts = 10 ** CODE_LENGTH
+    let payload = null
+
+    do {
+      payload = `${Math.round(Math.random() * (10 ** CODE_LENGTH))}`
+        .slice(0, CODE_LENGTH)
```

Comment:
```suggestion
```

Reply:
Would like to keep this as `Math.random()`. It may return `1` and produce `CODE_LENGTH` + 1 digits. Padding won't clip numbers.

```js
> `${Math.random() * (10 ** 6)}`.slice(0, 6).padStart(6, 0)
'892769'
> `${1 * (10 ** 6)}`.slice(0, 6).padStart(6, 0)
'100000'
> `${Math.random() * (10 ** 6)}`.padStart(6, 0)
'673722'
> `${1 * (10 ** 6)}`.padStart(6, 0)
'1000000'
       ^------- too long
```